core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -93,8 +93,29 @@ class LogSegment private[log] (val log: FileRecords,
   private var rollingBasedTimestamp: Option[Long] = None
 
   /* The maximum timestamp we see so far */
-  @volatile private var maxTimestampSoFar: Long = timeIndex.lastEntry.timestamp
-  @volatile private var offsetOfMaxTimestamp: Long = timeIndex.lastEntry.offset
+  @volatile private var maxTimestampSoFar: Long = _
+  @volatile private var offsetOfMaxTimestamp: Long = _
+
+
+  locally {
+    try {
+      /* For some cases (like KAFKA-6264), it might not be enough to lookup the last entry in time-index because we could
+       * have messages that were appended after time index was last updated. So lookup the last entry and scan messages
+       * that appear after that to make sure we have the correct maximum timestamp and its corresponding offset.
+       * Calling loadLargestTimestamp should not have a performance impact for the normal case where the last entry
+       * was appended successfully during shutdown.
+       */
+      loadLargestTimestamp()
+    } catch {
+      /* Log layer is responsible for initiating recovery of a segment, so ignore the exceptions here */
+      case _: CorruptRecordException |
+           _: CorruptIndexException |
+           _: NoSuchFileException => {
+        maxTimestampSoFar = timeIndex.lastEntry.timestamp
```

Comment:
Feels a tad odd to use the index entry if the index is corrupt even if we expect `Log` to fix the problem. Perhaps we could move this logic into an `initialize` function and let it raise the exceptions? We could then call it from `Log.loadSegmentFiles` like this:

```scala
try {
 segment.sanityCheck(timeIndexFileNewlyCreated)
 segment.initialize()
} catch {
...
```
Actually we could even move `sanityCheck` into `initialize` to simplify things a bit. The contract would then be that you have to call `segment.recover` if `segment.initialize` fails for any reason.

Reply:
Makes sense - will make the change.